File: mdlink/src/client/py/jzs/MdlinkClient.py
import sys
#sys.path.append( './jzsapi')
import jzs

import zmq
import time
import datetime
import threading
import traceback


class MdlinkClient (threading.Thread):

    def __init__(self, addr, callback) :
        threading.Thread.__init__(self)
        self._addr = addr
        self._callback = callback
        pass
    
    def run(self):
        ctx = zmq.Context()  
        sock = ctx.socket(zmq.SUB)
        sock = ctx.socket(zmq.SUB)
        sock.setsockopt(zmq.SUBSCRIBE, '')
        sock.connect(self._addr)
        
        while True:
            try:
                data = sock.recv()
                if not data:
                    time.sleep(0.1)
                    continue
            except :
                traceback.print_exc()
                time.sleep(0.1)
                continue
            msg = jzs.jzs_pb2.Msg()
            msg.ParseFromString(data[4:])
    
            if msg.head.tid != jzs.jzs_pb2.MSG_MD_MARKETDATA_IND:
                print "Unexpect msg type ", msg.head.tid
                continue

            ind = jzs.md_pb2.MarketDataInd()
            ind.ParseFromString(msg.body)
            
            if self._callback:
                self._callback(ind)
            else:
                #print ind
                if ind.HasField('fut'):
                    print ind.fut.jzcode, ind.fut.last
                elif ind.HasField('stk'):
                    print ind.stk.jzcode, ind.stk.last

        self._ctx = context
        self._sock = socket
        self._dst_node = dst_node
        
def test():
    mdlink = MdlinkClient('tcp://127.0.0.1:8700', None)
    mdlink.start()
    mdlink.join()

if __name__ == '__main__':
    test()

File: mdlink/src/client/jztsapi/src/main/java/jzs/api/MdlinkClient.java
package jzs.api;

import jzs.msg.*;
import jzs.msg.md.Md;
import jzs.msg.md.Md.MarketDataInd;

import java.util.Arrays;

import org.zeromq.*;
import org.zeromq.ZMQ;

public class MdlinkClient extends java.lang.Thread {

	public interface Callback {
		public void onMarketDataInd(Md.MarketDataInd ind);
	}

	String addr = "";
	Callback cb = null;

	public MdlinkClient(String addr, Callback cb) {
		this.addr = addr;
		this.cb = cb;
	}
	
	public void run(){
		ZContext ctx = new ZContext();
		
		ZMQ.Socket sock = ctx.createSocket(ZMQ.SUB);
		byte[] tid = new byte[4];
		for (int i=0; i < 4; i++) {
			tid[i] = (byte)((Jzs.MsgType.MSG_MD_MARKETDATA_IND_VALUE >> (i*8)));
		}
		sock.subscribe(new byte[0]);//tid);
		sock.connect(this.addr);
		
		while (true) {
			try {
				byte[] data = sock.recv();
				if ( data == null || data.length == 0) {
					Thread.sleep(10);
					continue;
				}
				
				byte[] tmp = Arrays.copyOfRange(data, 4, data.length);
				Jzs.Msg msg = Jzs.Msg.parseFrom(tmp);
				
				if (msg==null)
					continue;
				
				if (msg.getHead().getTid() != Jzs.MsgType.MSG_MD_MARKETDATA_IND_VALUE)
					continue;

				Md.MarketDataInd ind = Md.MarketDataInd.parseFrom(msg.getBody());
				if (ind != null && this.cb!=null) {
					this.cb.onMarketDataInd(ind);
				}
					
			}catch(Exception e){
				e.printStackTrace();
			}
		}
	}
	
	
	public static void Test(String addr) {
        MdlinkClient mdlink = new MdlinkClient(addr, new MdlinkClient.Callback() {

            @Override
            public void onMarketDataInd(MarketDataInd ind) {
                System.out.println(ind);
            }
            
        });
        mdlink.start();
    }
    
    static public void main(String[] args) throws InterruptedException{
        Test("tcp://127.0.0.1:8700");
    }
}
